src/vsr/journal.zig

```diff
@@ -669,117 +812,541 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+                    assert(prepare.command == .reserved);
+                    assert(!self.dirty.bit(slot));
+                    assert(!self.faulty.bit(slot));
+                    assert(self.prepare_checksums[slot.index] == 0);
+                },
+                .fix => {
+                    assert(prepare.command == .prepare);
+                    assert(self.prepare_checksums[slot.index] == prepare.checksum);
+                    if (replica.replica_count == 1) {
+                        self.headers[slot.index] = prepare.*;
+                        self.dirty.clear(slot);
+                    } else {
+                        // TODO Repair without retrieving remotely (i.e. don't set dirty or faulty).
+                        // This is tricky because the redundant headers are written in batches.
+                        self.set_header_as_dirty(prepare);
+                    }
+                    self.faulty.clear(slot);
+                },
+                .vsr => {
+                    // Faulty headers are loaded as reserved to prevent unsafe batch updates of the
+                    // redundant headers. See `recover_prepares_callback()` for more detail.
+                    self.headers[slot.index] = Header.reserved(replica.cluster, slot.index);
+                    self.dirty.set(slot);
+                    self.faulty.set(slot);
+                },
+            }
+
+            switch (case.decision) {
+                .eql, .nil => {
+                    log.debug("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
+                .fix, .vsr => {
+                    log.warn("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
             }
 
-            // We must release before we call `recover_headers()` in case Storage is synchronous.
-            // Otherwise, we would run out of messages and reads.
             replica.message_bus.unref(read.message);
             self.reads.release(read);
+            self.recover_prepares(Slot{ .index = slot.index + 1 });
+        }
+
+        fn recover_done(self: *Self) void {
+            const replica = @fieldParentPtr(Replica, "journal", self);
+
+            log.debug("{}: recover_done: dirty={} faulty={}", .{
+                self.replica,
+                self.dirty.count,
+                self.faulty.count,
+            });
+
+            assert(self.faulty.count == 0 or replica.replica_count != 1);
+            assert(self.faulty.count >= self.dirty.count);
+            assert(self.faulty.count < slot_count);
```

Comment:
Actually this is deliberate (though I can add a note) — if every slot is faulty, something is Very Wrong and we abort.

Reply:
That makes sense.

Let's also do both:

1. Leave the assert as showing the invariant possibility that all slots can be faulty, since that's within the positive space, and could become more probable if `config.journal_slot_count` is set low, e.g. for testing purposes.

2. And then also deal with this as an explicit *policy decision*, i.e. to shutdown if things get that bad with an explicit `@panic()`.